### ISSUE TITLE
Add button method that works like link tag (accepts route helper)

### ISSUE THREAD

**Issue by paulcsmith** [luckyframework/web#100]
See the link helper and link helper spec (https://github.com/luckyframework/web/blob/master/src/lucky_web/tags/link_helpers.cr). It will work very similarly, but will generate a `button` tag instead

Since Crystal supports method overloading I *think* we can just call this `button` and it will work since it accepts a `LuckyWeb::RouteHelper` and therefore will not overwrite the regular `button` method for when you don't want to pass it a route helper

**Comment by 5minpause** [comment]
I am trying to do this as well.

**Comment by 5minpause** [comment]
Alright, this might prove too much for my non-existing Crystal skills.

What did I do:
1. create a new file `lucky_web/tags/button_helpers.rc`
```crystal
require "./link_helpers"

module LuckyWeb::ButtonHelpers
  include LuckyWeb::LinkHelpers

  def button(text, to : LuckyWeb::RouteHelper, **html_options)
    a text, merge_options(html_options, link_to_href(to))
  end

  def button(text, to : LuckyWeb::Action.class, **html_options)
    a text, merge_options(html_options, link_to_href(to.route))
  end

  def button(to : LuckyWeb::RouteHelper, **html_options)
    a merge_options(html_options, link_to_href(to)) do
      yield
    end
  end

  def button(to : LuckyWeb::Action.class, **html_options)
    a merge_options(html_options, link_to_href(to.route)) do
      yield
    end
  end

  def button(text, to : String, **html_options)
    a text, merge_options(html_options, {"href" => to})
  end

  def button(to : String, **html_options)
    a merge_options(html_options, {"href" => to}) do
      yield
    end
  end
end
```
2. create a new spec file `spec/lucky_web/button_helpers_spec.rb`
```crystal
require "../../spec_helper"

class ButtonHelpers::Index < LuckyWeb::Action
  action { render_text "foo" }
end

class ButtonHelpers::Create < LuckyWeb::Action
  action { render_text "foo" }
end

private class TestPage
  include LuckyWeb::Page

  render do
  end

  def get_route
    button "Test", to: ButtonHelpers::Index.route
  end

  def non_get_route
    button "Test", to: ButtonHelpers::Create.route
  end

  def non_get_route_with_options
    button "Test", to: ButtonHelpers::Create.route, something_custom: "foo"
  end

  def string_path
    button "Test", to: "/foos"
  end

  def string_path_with_options
    button "Test", to: "/foos", data_method: "post"
  end

  def get_route_with_block
    button to: ButtonHelpers::Index.route do
      text "Hello"
    end
  end

  def string_path_with_block
    button to: "/foo" do
      text "Hello"
    end
  end
end

describe LuckyWeb::ButtonHelpers do
  #   it "renders a button tag" do
  #     view.get_route.to_s.should contain %(<a href="/link_helpers">Test</a>)
  #     view.non_get_route.to_s.should contain %(<a href="/link_helpers" data-method="post">Test</a>)
  #     view
  #       .non_get_route_with_options
  #       .to_s
  #       .should contain %(<a href="/link_helpers" data-method="post" something-custom="foo">Test</a>)
  #     view.string_path.to_s.should contain %(<a href="/foos">Test</a>)
  #     view.string_path_with_options.to_s.should contain %(<a href="/foos" data-method="post">Test</a>)
  #   end

  #   it "renders a link tag with an action" do
  #     view.link("Test", to: ButtonHelpers::Index).to_s.should contain <<-HTML
  #     <a href="/link_helpers">Test</a>
  #     HTML

  #     link = view.link(to: ButtonHelpers::Index, class: "link") { }

  #     link.to_s.should contain <<-HTML
  #     <a href="/link_helpers" class="link"></a>
  #     HTML
  #   end

  it "renders a link tag with a block" do
    view.string_path_with_block.to_s.should contain <<-HTML
    <a href="/foo">Hello</a>
    HTML

    view.get_route_with_block.to_s.should contain <<-HTML
    <a href="/link_helpers">Hello</a>
    HTML
  end
end

private def view
  TestPage.new
end
```

The Overloading does not seem to work like it should? I am probably doing something wrong and it might work…
```
± crystal spec                                                                                                                                          2.3.3
Error in line 12: while requiring "./spec/lucky_web/button_helpers_spec.cr"

in spec/lucky_web/button_helpers_spec.cr:79: instantiating 'TestPage#get_route_with_block()'

    view.get_route_with_block.to_s.should contain <<-HTML
         ^~~~~~~~~~~~~~~~~~~~

in spec/lucky_web/button_helpers_spec.cr:38: instantiating 'button()'

    button to: ButtonHelpers::Index.route do
    ^~~~~~

in macro 'macro_4515558160' /Users/holger/projects/ifeellucky/web/src/lucky_web/tags/base_tags.cr:38, line 8:

   1.
   2.     generate_tag_methods(method_name: a, tag: a)
   3.
   4.     generate_tag_methods(method_name: b, tag: b)
   5.
   6.     generate_tag_methods(method_name: body, tag: body)
   7.
>  8.     generate_tag_methods(method_name: button, tag: button)
   9.
  10.     generate_tag_methods(method_name: div, tag: div)
  11.
  12.     generate_tag_methods(method_name: em, tag: em)
  13.
  14.     generate_tag_methods(method_name: small, tag: small)
  15.
  16.     generate_tag_methods(method_name: fieldset, tag: fieldset)
  17.
  18.     generate_tag_methods(method_name: h1, tag: h1)
  19.
  20.     generate_tag_methods(method_name: h2, tag: h2)
  21.
  22.     generate_tag_methods(method_name: h3, tag: h3)
  23.
  24.     generate_tag_methods(method_name: h4, tag: h4)
  25.
  26.     generate_tag_methods(method_name: h5, tag: h5)
  27.
  28.     generate_tag_methods(method_name: h6, tag: h6)
  29.
  30.     generate_tag_methods(method_name: head, tag: head)
  31.
  32.     generate_tag_methods(method_name: html, tag: html)
  33.
  34.     generate_tag_methods(method_name: i, tag: i)
  35.
  36.     generate_tag_methods(method_name: label, tag: label)
  37.
  38.     generate_tag_methods(method_name: li, tag: li)
  39.
  40.     generate_tag_methods(method_name: ol, tag: ol)
  41.
  42.     generate_tag_methods(method_name: option, tag: option)
  43.
  44.     generate_tag_methods(method_name: s, tag: s)
  45.
  46.     generate_tag_methods(method_name: script, tag: script)
  47.
  48.     generate_tag_methods(method_name: span, tag: span)
  49.
  50.     generate_tag_methods(method_name: strong, tag: strong)
  51.
  52.     generate_tag_methods(method_name: table, tag: table)
  53.
  54.     generate_tag_methods(method_name: tbody, tag: tbody)
  55.
  56.     generate_tag_methods(method_name: td, tag: td)
  57.
  58.     generate_tag_methods(method_name: textarea, tag: textarea)
  59.
  60.     generate_tag_methods(method_name: thead, tag: thead)
  61.
  62.     generate_tag_methods(method_name: th, tag: th)
  63.
  64.     generate_tag_methods(method_name: title, tag: title)
  65.
  66.     generate_tag_methods(method_name: tr, tag: tr)
  67.
  68.     generate_tag_methods(method_name: u, tag: u)
  69.
  70.     generate_tag_methods(method_name: ul, tag: ul)
  71.
  72.     generate_tag_methods(method_name: form, tag: form)
  73.
  74.     generate_tag_methods(method_name: footer, tag: footer)
  75.
  76.     generate_tag_methods(method_name: header, tag: header)
  77.
  78.     generate_tag_methods(method_name: article, tag: article)
  79.
  80.     generate_tag_methods(method_name: aside, tag: aside)
  81.
  82.     generate_tag_methods(method_name: bdi, tag: bdi)
  83.
  84.     generate_tag_methods(method_name: details, tag: details)
  85.
  86.     generate_tag_methods(method_name: dialog, tag: dialog)
  87.
  88.     generate_tag_methods(method_name: figcaption, tag: figcaption)
  89.
  90.     generate_tag_methods(method_name: figure, tag: figure)
  91.
  92.     generate_tag_methods(method_name: main, tag: main)
  93.
  94.     generate_tag_methods(method_name: mark, tag: mark)
  95.
  96.     generate_tag_methods(method_name: menuitem, tag: menuitem)
  97.
  98.     generate_tag_methods(method_name: meter, tag: meter)
  99.
 100.     generate_tag_methods(method_name: nav, tag: nav)
 101.
 102.     generate_tag_methods(method_name: progress, tag: progress)
 103.
 104.     generate_tag_methods(method_name: rp, tag: rp)
 105.
 106.     generate_tag_methods(method_name: rt, tag: rt)
 107.
 108.     generate_tag_methods(method_name: ruby, tag: ruby)
 109.
 110.     generate_tag_methods(method_name: section, tag: section)
 111.
 112.     generate_tag_methods(method_name: summary, tag: summary)
 113.
 114.     generate_tag_methods(method_name: time, tag: time)
 115.
 116.     generate_tag_methods(method_name: wbr, tag: wbr)
 117.

expanding macro
in macro 'generate_tag_methods' /Users/holger/projects/ifeellucky/web/src/lucky_web/tags/base_tags.cr:9, line 21:

   1.     def button(content = "", options = EMPTY_HTML_ATTRS, **other_options)
   2.       merged_options = merge_options(other_options, options)
   3.       button(merged_options) do
   4.         text content
   5.       end
   6.     end
   7.
   8.     def button(content : String)
   9.       button(EMPTY_HTML_ATTRS) do
  10.         text content
  11.       end
  12.     end
  13.
  14.     def button(&block)
  15.       button(EMPTY_HTML_ATTRS) do
  16.         yield
  17.       end
  18.     end
  19.
  20.     def button(options = EMPTY_HTML_ATTRS, **other_options, &block)
> 21.       merged_options = merge_options(other_options, options)
  22.       tag_attrs = build_tag_attrs(merged_options)
  23.       @view << "<button" << tag_attrs << ">"
  24.       yield
  25.       @view << "</button>"
  26.     end
  27.

instantiating 'merge_options(NamedTuple(to: LuckyWeb::RouteHelper), Hash(String, String))'
in src/lucky_web/tags/base_tags.cr:78: instantiating 'NamedTuple(to: LuckyWeb::RouteHelper)#each()'

    html_options.each do |key, value|
                 ^~~~

in src/lucky_web/tags/base_tags.cr:78: instantiating 'NamedTuple(to: LuckyWeb::RouteHelper)#each()'

    html_options.each do |key, value|
                 ^~~~

in src/lucky_web/tags/base_tags.cr:79: no overload matches 'Hash(String, String)#[]=' with types String, LuckyWeb::RouteHelper
Overloads are:
 - Hash(K, V)#[]=(key : K, value : V)

      options[key.to_s] = value
             ^

```

**Comment by paulcsmith** [comment]
I think the problem may be that there is a default `button` tag generated in `BaseTags`

https://github.com/luckyframework/web/blob/master/src/lucky_web/tags/base_tags.cr#L2

So maybe you need to remove `button` from that list first?

**Comment by 5minpause** [comment]
That makes sense. I noticed that it was included there, but did not know how to proceed. Thank you.

I removed it and added the tests for the ButtonHelpers module. It is mostly a copy and paste of the link helpers spec file. Is there a logical way to exclude the duplication? Maybe define the `private class TestPage` in a `/support/*` file and require/include it here? One could parameterize the methods inside the class and then use it depending on whether we use it inside the Link or Button Helpers.
Does that make sense? does that easily work in Crystal?
